controllers/vmware/vspherecluster_reconciler.go

```diff
@@ -121,6 +121,13 @@ func (r ClusterReconciler) Reconcile(_ goctx.Context, req ctrl.Request) (_ ctrl.
 		return reconcile.Result{}, nil
 	}
 
+	// If the VSphereCluster doesn't have our finalizer, add it.
+	// Requeue immediately after adding finalizer to avoid the race condition between init and delete
+	if !controllerutil.ContainsFinalizer(vsphereCluster, vmwarev1.ClusterFinalizer) {
```

Comment:
Nit/optional: we could move this up a bit, above the `EmitEvent` so "just adding the finalizer" does not emit an event. (Note: we would have to add `vsphereCluster.DeletionTimestamp.IsZero()` to the if, if we are above the deletion part)

Reply:
I think let's leave it here. we need the defer patch to set the finalizer